### PR TITLE
Try to fix dropped data in P2P shuffle

### PR DESCRIPTION
**Pull request by gjoseph92**
Attempt with @graingert at fixing #6277; doesn't work.

Also fixes a couple other small design issues (`ShuffleExtension` peeking into implementation details of `Shuffle` during `shuffle_receive`/`shuffle_inputs_done`.)

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/shuffle/shuffle_extension.py [hunk 1]
```diff
@@ -324,7 +350,10 @@ def get_output_partition(
             shuffle = self.shuffles.pop(shuffle_id, None)
             # key missing if another thread got to it first
             if shuffle:
-                shuffle.close()
+                sync(
```
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/shuffle/shuffle_extension.py:353]: Agreed, I was just lazy here

**Comment by graingert** [comment]
see https://github.com/dask/distributed/issues/6279

**Comment by fjetter** [comment]
Next time something like this happens, please provide a bit of information why it doesn't work. A reproducer of the issue would also be welcome.

**Comment by hendrikmakait** [comment]
This should be superseded by #7486 and previous changes.

**Pull request closed without merging**